FirebaseStorage/Tests/SwiftIntegration/StorageIntegration.swift

```diff
@@ -0,0 +1,657 @@
+
+    _ = [ref?.putData(data, metadata: nil, completion: { metadata, error in
+      XCTAssertNotNil(metadata, "Metadata should not be nil")
+      XCTAssertNil(error, "Error should be nil")
+      let task = ref?.write(toFile: fileURL)
+
+      task?.observe(StorageTaskStatus.success, handler: { snapshot in
+        do {
+          let stringData = try String(contentsOf: fileURL, encoding: .utf8)
+          XCTAssertEqual(stringData, "Hello Swift World")
+          XCTAssertEqual(snapshot.description, "<State: Success>")
+          expectation.fulfill()
+        } catch {
+          XCTFail("Exception processing success snapshot")
+        }
+      })
+
+      task?.observe(StorageTaskStatus.progress, handler: { snapshot in
+        XCTAssertNil(snapshot.error, "Error should be nil")
+        guard let progress = snapshot.progress else {
+          XCTFail("Missing progress")
+          return
+        }
+        print("\(progress.completedUnitCount) of \(progress.totalUnitCount)")
+      })
+      task?.observe(StorageTaskStatus.failure, handler: { snapshot in
+        XCTAssertNil(snapshot.error, "Error should be nil")
+      })
+    })]
+    waitForExpectations()
+  }
+
+  func testCancelDownload() throws {
+    let expectation = self.expectation(description: "testCancelDownload")
+    let ref = storage?.reference(withPath: "ios/public/1mb")
+    let tmpDirURL = URL(fileURLWithPath: NSTemporaryDirectory())
+    let fileURL = tmpDirURL.appendingPathComponent("hello.dat")
+
+    let task = ref?.write(toFile: fileURL)
+
+    task?.observe(StorageTaskStatus.failure, handler: { snapshot in
+      XCTAssertTrue(snapshot.description.starts(with: "<State: Failed"))
+      expectation.fulfill()
+    })
+
+    task?.observe(StorageTaskStatus.progress, handler: { _ in
+      task?.cancel()
+    })
+    waitForExpectations()
+  }
+
+  private func assertMetadata(actualMetadata: StorageMetadata,
+                              expectedContentType: String,
+                              expectedCustomMetadata: [String: String]) {
+    XCTAssertEqual(actualMetadata.cacheControl, "cache-control")
+    XCTAssertEqual(actualMetadata.contentDisposition, "content-disposition")
+    XCTAssertEqual(actualMetadata.contentEncoding, "gzip")
+    XCTAssertEqual(actualMetadata.contentLanguage, "de")
+    XCTAssertEqual(actualMetadata.contentType, expectedContentType)
+    XCTAssertTrue(actualMetadata.md5Hash!.count == 24)
+    for (key, value) in expectedCustomMetadata {
+      XCTAssertEqual(actualMetadata.customMetadata![key], value)
+    }
+  }
+
+  private func assertMetadataNil(actualMetadata: StorageMetadata) {
+    XCTAssertNil(actualMetadata.cacheControl)
+    XCTAssertNil(actualMetadata.contentDisposition)
+    XCTAssertEqual(actualMetadata.contentEncoding, "identity")
+    XCTAssertNil(actualMetadata.contentLanguage)
+    XCTAssertNil(actualMetadata.contentType)
+    XCTAssertTrue(actualMetadata.md5Hash!.count == 24)
```

Comment:
Nit: `XCTAssertEqual(actualMetadata.md5Hash?.count, 24)` would produce a bit more informative error in case of failure.

Reply:
Done